plugins/opds.koplugin/opdspse.lua

```diff
@@ -0,0 +1,217 @@
+            method = "POST",
+            url         = auth_url,
+            headers     = {
+                ["Accept-Encoding"] = "identity",
+                ["Authentication"] = api_key,
+            },
+            sink        = ltn12.sink.table(auth_data),
+            user        = self.root_catalog_username,
+            password    = self.root_catalog_password,
+        })
+        socketutil:reset_timeout()
+    else
+        UIManager:show(InfoMessage:new {
+            text = T(_("Invalid protocol:\n%1"), auth_parsed.scheme),
+        })
+    end
+
+    if auth_code == 200 then
+        -- if http request for bearer token was successful, pull bearer token from response and
+        -- attempt to pull progress for chapterId in remote_url
+        local bearer_token = auth_data[1]:match("\"token\":\"(.+)\",\"refresh")
+
+        -- Do HTTP GET request for chapter progress
+        local progress_parsed = url.parse(progress_url)
+        local progress_data = {}
+        local progress_code, progress_headers, progress_status
+        if progress_parsed.scheme == "http" or progress_parsed.scheme == "https" then
+            socketutil:set_timeout(socketutil.FILE_BLOCK_TIMEOUT, socketutil.FILE_TOTAL_TIMEOUT)
+            progress_code, progress_headers, progress_status = socket.skip(1, http.request {
+                url         = progress_url,
+                headers     = {
+                    ["Accept-Encoding"] = "identity",
+                    ["Authorization"] = "Bearer "..bearer_token,
+                },
+                sink        = ltn12.sink.table(progress_data),
+                user        = self.root_catalog_username,
+                password    = self.root_catalog_password,
+            })
+            socketutil:reset_timeout()
+        else
+            UIManager:show(InfoMessage:new {
+                text = T(_("Invalid protocol:\n%1"), progress_parsed.scheme),
+            })
+        end
+
+        if progress_code == 200 then
+            -- if HTTP GET was successful, pull page number from response
+            last_page = progress_data[1]:match("\"pageNum\":(.+),\"seriesId")
+        else
+            logger.dbg("OPDSPSE:getLastPage: Progress Request failed:", progress_status or progress_code)
+            logger.dbg("OPDSPSE:getLastPage: Progress Response headers:", progress_headers)
+        end
+    else
+        logger.dbg("OPDSPSE:getLastPage: Authentication Request failed:", auth_status or auth_code)
+        logger.dbg("OPDSPSE:getLastPage: Authentication Response headers:", auth_headers)
+    end
+
+    -- returns page number. If the HTTP Requests were unsuccessful, defaults to 0.
+    return last_page;
+end
+
+
+
+function OPDSPSE:streamPages(item, remote_url, count, continue)
+    -- attempt to pull chapter progress from Kavita if user pressed
+    -- "Page Stream" button
+    local ok, last_page = pcall(function() return self:getLastPage(remote_url) end)

```

Comment:
Let's move getting `last_page` to the place where it is used.

Reply:
It needs to be at the top of the function, before the page_table is created. The creation of page_table resets your book progress in Kavita, so the progress has to be pulled before the page_table is created.

Reply:
Okay, didn't know that.